src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java

```diff
@@ -992,6 +1021,35 @@ private void rollbackTransaction(){
             }
         }
 
+        public void addUpdateToPersistence(String label, byte[] data){
+            addMessageToPersistence(label, PERSISTENCE_MAP_MQTT_UPDATE, updatePersistenceLimit, data);
+        }
+
+        public void addLogToPersistence(String label, byte[] data){
+            addMessageToPersistence(label, PERSISTENCE_MAP_MQTT_LOG, logPersistenceLimit, data);
+        }
+
+        private void addMessageToPersistence(String label, String MapKey, int persistenceLimit, byte[] data){
+            if(persistenceLimit<0) {
+                synchronized (persistenceLock) {
+                    DB dbPersistenceData = DBMaker.fileDB(PERSISTENCE_FILE).make();
```

Comment:
Persistence database is opened and closed every time a message has to be added. Could we not open the databases when initialising the endpoint and use them as attributes without closing and re-opening them every time. As a plus we could use the database objects for the mutex section.

Another idea would be to put all persistence functionality into a separate class.

Reply:
> Persistence database is opened and closed every time a message has to be added. Could we not open the databases when initialising the endpoint and use them as attributes without closing and re-opening them every time. As a plus we could use the database objects for the mutex section.

The only way I found we could avoid closing the database all the time would be to use the transaction of mapdb https://jankotek.gitbooks.io/mapdb/content/db/ . Does it make sense to you?

> Another idea would be to put all persistence functionality into a separate class.

I think it will be cumbersome to do that especially when we empty the maps and to send mqtt messages.

Reply:
> The only way I found we could avoid closing the database all the time would be to use the transaction of mapdb https://jankotek.gitbooks.io/mapdb/content/db/ . Does it make sense to you?

Why that? I mean the access to the database is protected by a mutex, so I see no reason there could be a race condition or it that mandatory to keep the size at a certain limit?

> I think it will be cumbersome to do that especially when we empty the maps and to send mqtt messages.

Basically the idea of a separate class for persistent storage is to use an interface for all persistent storage operations from the point of view of the endpoint and then providing a default implementation based on mapdb. This would allow a developer to provide his own persistence layer that might use a completely different technology to save the persistent data.

Reply:
> Why that? I mean the access to the database is protected by a mutex, so I see no reason there could be a race condition or it that mandatory to keep the size at a certain limit?

Thing is if we don't do a proper db.close() the db get corrupted easily and we can't access it anymore. So only solution I "found" is to do it with transaction and to use db.commit to be sure the data are saved on the db. 

> Basically the idea of a separate class for persistent storage is to use an interface for all persistent storage operations from the point of view of the endpoint and then providing a default implementation based on mapdb.

It's more clear now, I agree with that.

Reply:
> Thing is if we don't do a proper db.close() the db get corrupted easily and we can't access it anymore. So only solution I "found" is to do it with transaction and to use db.commit to be sure the data are saved on the db.

Strange. Well if the transactions solves the problem then we should use them.